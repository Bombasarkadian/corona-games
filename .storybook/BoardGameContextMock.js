import React from "react";
import { boolean, number, text, select } from "@storybook/addon-knobs";
import { action } from "@storybook/addon-actions";
import { BoardGameProvider } from "contexts/BoardGameContext";

const ref = "Game Context";

const BoardGameContextMock = ({ children }) => {
  return (
    <BoardGameProvider
      G={{
        timePerTurn: 120,
        turnEndTime:
          Math.floor(new Date().getTime() / 1000) +
          number(
            "Remaining seconds",
            100,
            {
              range: true,
              min: 0,
              max: 120,
              step: 1,
            },
            ref
          ),
        points: [10, 3],
        actions: [
          { action: "message", text: text("Message text", "Hello there, shall we begin?", ref) },
        ],
        privateRoom: boolean("Private", true, ref),
        players: {
          "0": {
            phrase: text("Phrase", "Baba z wozu", ref),
            card: { pictures: [], layout: 0, rotation: 0 },
          },
          "1": {},
        },
        canChangePhrase: boolean("Can change phrase?", true, ref),
        currentCard: { pictures: [], layout: 0, rotation: 0 },
      }}
      ctx={{
        activePlayers: [
          select("stage", ["draw", "guess", "wait", "manage", "match"], "draw", ref),
          "guess",
        ],
        phase: select("phase", ["wait", "play"], "play", ref),
        numPlayers: number("Player count", 4, { range: true, min: 2, max: 10, step: 1 }, ref),
      }}
      moves={{
        Guess: action("Guess"),
        Ping: action("Ping"),
        UpdateDrawing: action("UpdateDrawing"),
        NotifyTimeout: action("NotifyTimeout"),
      }}
      playerID={0}
      gameName={"Kalambury"}
      gameID={"qwe123"}
      rawClient={{ transport: { socket: null } }}
      gameMetadata={[
        { id: 0, name: "user-0", isConnected: true },
        { id: 1, name: "user-1", isConnected: false },
      ]}
    >
      {children}
    </BoardGameProvider>
  );
};

export default BoardGameContextMock;
