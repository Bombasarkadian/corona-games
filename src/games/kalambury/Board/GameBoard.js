import React, { useCallback, useEffect, useState, useMemo } from "react";
import Confetti from "react-dom-confetti";

import { currentTime } from "../utils/time";
import { WIDE_CONFETTI } from "config/confetti";
import { useBoardGame } from "contexts/BoardGameContext";
import filterActions from "utils/user/filterActions";
import { throttleAccumulate } from "utils/throttle";

import DrawArea from "../DrawArea";
import GuessingBoard from "./GuessingBoard";
import Countdown from "./Countdown";

const GameBoard = ({ guess, setGuess, envokeLastAnswer, guessInputRef }) => {
  const {
    G,
    moves,
    player: { isDrawing, phrase },
    playerID,
    chatMessages,
    sendChatMessage,
  } = useBoardGame();
  const [lines, setLines] = useState([]);
  const [remainingSeconds, setRemainingSeconds] = useState(G.turnEndTime - currentTime());
  const [lastSuccess, setLastSuccess] = useState(false);
  const lastUserGuess = filterActions(G.actions, playerID, "guess")[0] || {
    id: null,
    success: false,
  };

  const throttledSendAppend = useMemo(
    () =>
      throttleAccumulate(sendChatMessage, 200, ([acc], [args]) => [
        {
          type: acc.type,
          data: [...acc.data, ...args.data],
        },
      ]),
    [sendChatMessage]
  );

  // send full drawing every 1sec
  useEffect(() => {
    if (isDrawing) {
      const interval = setInterval(() => {
        throttledSendAppend.flush();
        sendChatMessage({ type: "UpdateDrawing:replace", data: lines });
      }, 1000);
      return () => clearInterval(interval);
    }
  }, [isDrawing, lines, sendChatMessage, throttledSendAppend]);

  // countdown timer
  useEffect(() => {
    const interval = setInterval(() => {
      const seconds = G.turnEndTime - currentTime();
      if ((seconds <= 0 && isDrawing) || seconds < -5) {
        moves.NotifyTimeout();
      }
      setRemainingSeconds(Math.max(seconds, 0));
    }, 1000);
    return () => clearInterval(interval);
  }, [G.turnEndTime, setRemainingSeconds, moves, isDrawing]);

  useEffect(() => {
    setLastSuccess(lastUserGuess.success);
    return () => setLastSuccess(false); // Do cleanup of state
  }, [lastUserGuess.id, lastUserGuess.success]);

  useEffect(() => {
    if (isDrawing) {
      setLines([]);
    }
  }, [isDrawing, phrase]);

  const updateLines = useCallback(
    (type, data = null) => {
      setLines((lines) => {
        if (type === "add" && lines) {
          return [...lines, data];
        }
        if (type === "append" && lines.length > 0) {
          const last = lines[lines.length - 1];
          return [...lines.slice(0, -1), { ...last, points: [...last.points, ...data] }];
        }
        if (type === "delete" && lines.length > 0) {
          return lines.slice(0, -1);
        }
        if (type === "replace" && data) {
          return data;
        }
        return lines;
      });
    },
    [setLines]
  );

  useEffect(() => {
    if (isDrawing || chatMessages.length <= 0) return;
    const lastMessage = chatMessages[chatMessages.length - 1].payload;

    if (lastMessage.type && lastMessage.type.startsWith("UpdateDrawing")) {
      const updateType = lastMessage.type.split(":")[1];
      updateLines(updateType, lastMessage.data);
    }
  }, [isDrawing, chatMessages, updateLines]);

  const handleLinesUpdate = useCallback(
    (type, data) => {
      updateLines(type, data);
      if (type === "append") {
        throttledSendAppend({ type: `UpdateDrawing:${type}`, data });
      } else {
        throttledSendAppend.flush();
        sendChatMessage({ type: `UpdateDrawing:${type}`, data });
      }
    },
    [updateLines, sendChatMessage, throttledSendAppend]
  );

  return (
    <>
      {isDrawing ? (
        <DrawArea lines={lines} updateLines={handleLinesUpdate} />
      ) : (
        <GuessingBoard
          lastUserGuess={lastUserGuess}
          envokeLastAnswer={envokeLastAnswer}
          guessInputRef={guessInputRef}
          guess={guess}
          setGuess={setGuess}
          lines={lines}
          remainingSeconds={remainingSeconds}
        />
      )}
      <Countdown remainingSeconds={remainingSeconds} totalTime={G.timePerTurn} />
      <Confetti active={lastSuccess} config={WIDE_CONFETTI} className="confetti" />
    </>
  );
};

export default GameBoard;
