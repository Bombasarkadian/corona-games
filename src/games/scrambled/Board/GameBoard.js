import React, { useState, useEffect, useCallback, useRef } from "react";
import { useTranslation } from "react-i18next";
import { Header, Confirm, Sticky } from "semantic-ui-react";

import { useBoardGame } from "contexts/BoardGameContext";
import GameLayout from "components/layout/GameLayout";

import Grid from "./Grid";
import TilesPanel from "./TilesPanel";
import InfoPopup from "./InfoPopup";
import { remove, orderBy, pick } from "lodash";
import { availableLaguages } from "../data/tiles";
import { tilesPlacementErrors, newWords, filterPlayedTiles, canPlaceTile } from "../utils";
import ApprovalModal from "../Modals/ApprovalModal";
import SwapTilesModal from "../Modals/SwapTilesModal";
import BlankTileModal from "../Modals/BlankTileModal";

const GameBoard = () => {
  const {
    G,
    ctx: { phase, currentPlayer },
    player: { id, tiles, stage, isCurrentPlayer },
    moves: { PlayTiles, SwapTiles, SkipTurn, Approve },
    chatMessages,
    sendChatMessage,
  } = useBoardGame();

  const { t } = useTranslation("scrambled");
  const [selectedTile, setSelectedTile] = useState(null);
  const [playerTiles, setPlayerTiles] = useState([]);
  const [moveErrors, setMoveErrors] = useState([]);
  const [selectedBlankTile, setSelectedBlankTile] = useState(null);
  const [blankTileReplacement, setBlankTileReplacement] = useState(null);
  const [swapTilesModal, setSwapTilesModal] = useState(false);
  const [skipTurnModal, setSkipTurnModal] = useState(false);
  const [selectedForSwap, setSelectedForSwap] = useState([]);
  const [popupOpen, setPopupOpen] = useState(false);
  const [previewTiles, setPreviewTiles] = useState(null);
  const popupHandleRef = useRef();
  const stickyRef = useRef();

  const currentLanguage = availableLaguages.find(({ key }) => key === G.language);
  const playedTiles = filterPlayedTiles(playerTiles);
  const canMakeMove = isCurrentPlayer && phase === "play" && stage === "play";
  const canReturnTiles = playedTiles.length > 0;
  const canPlayTiles = playedTiles.length > 0 && moveErrors.length === 0;
  const canSwapTiles = G.tilesLeft >= 7;
  const approvalPhase = phase === "play" && stage === "approve";

  const onReturnTiles = useCallback(() => {
    Object.values(playerTiles).forEach((tile) => {
      delete tile.x;
      delete tile.y;
      delete tile.popupRef;
    });
    setPlayerTiles([...playerTiles]);
    setPopupOpen(false);
  }, [playerTiles]);

  const onPlayTiles = useCallback(() => {
    playerTiles.forEach((tile) => {
      tile.popupRef = undefined;
    });
    setPopupOpen(false);
    PlayTiles({ tiles: playerTiles });
  }, [PlayTiles, playerTiles]);

  const onSkipTurn = useCallback(() => {
    setSkipTurnModal(false);
    playerTiles.forEach((tile) => {
      tile.popupRef = undefined;
    });
    setPopupOpen(false);
    SkipTurn();
  }, [SkipTurn, playerTiles]);

  const onSelectForSwap = useCallback(
    (tile) => {
      selectedForSwap.includes(tile)
        ? remove(selectedForSwap, (el) => el === tile)
        : selectedForSwap.push(tile);
      setSelectedForSwap([...selectedForSwap]);
    },
    [selectedForSwap]
  );

  useEffect(() => {
    if (tiles !== undefined) setPlayerTiles(tiles);
  }, [tiles]);

  useEffect(() => {
    // player.isCurrentPlayer didn't seem to work properly
    if (id !== undefined && id.toString() !== currentPlayer) {
      setMoveErrors(["invalid_player"]);
      return;
    }
    const playedTiles = filterPlayedTiles(playerTiles);
    setMoveErrors(tilesPlacementErrors(G, currentPlayer, playedTiles));
  }, [G, currentPlayer, id, playerTiles]);

  useEffect(() => {
    if (canMakeMove || chatMessages.length <= 0) return;
    const lastMessage = chatMessages[chatMessages.length - 1].payload;

    if (lastMessage.type && lastMessage.type.startsWith("MovePreview")) {
      setPreviewTiles(lastMessage.data.map((el) => ({ preview: true, ...el })));
    }
  }, [canMakeMove, chatMessages]);

  const selectTile = useCallback(
    (tile) => {
      if (!canMakeMove) return;
      if (selectedTile === tile) setSelectedTile(null);
      else setSelectedTile(tile);
    },
    [canMakeMove, selectedTile]
  );

  const canBeClicked = useCallback(
    (x, y) => {
      // Player can't click outside of their turn
      if (!isCurrentPlayer) return false;
      // Player can always click on a temporarily placed tile to remove it
      if (playerTiles.find((tile) => tile.x === x && tile.y === y)) return true;
      if (selectedTile !== null) return canPlaceTile(G, playedTiles, x, y);
      return false;
    },
    [G, isCurrentPlayer, playedTiles, playerTiles, selectedTile]
  );

  const clickBoard = useCallback(
    (x, y) => {
      if (!canBeClicked(x, y)) return;
      setPreviewTiles(null);

      // Remove a tile that's on x,y
      const currentIdx = playerTiles.findIndex((tile) => tile.x === x && tile.y === y);
      if (currentIdx >= 0) {
        delete playerTiles[currentIdx].x;
        delete playerTiles[currentIdx].y;
      }

      // Place new tile on x,y
      if (selectedTile !== null) {
        if (!selectedTile.letter) {
          setSelectedBlankTile(selectedTile);
        }
        selectedTile.x = x;
        selectedTile.y = y;
      }

      setSelectedTile(null);
      // Force re-rendering of placed tiles
      setPlayerTiles([...playerTiles]);

      playerTiles.forEach((tile) => {
        tile.popupRef = undefined;
      });

      const playedTiles = filterPlayedTiles(playerTiles);
      // Select new tile for holding info popup
      if (playedTiles.length > 0)
        orderBy(playedTiles, ["y", "x"], ["asc", "asc"])[0].popupRef = popupHandleRef;
      setPopupOpen(playedTiles.length > 0);

      if (G.preview)
        sendChatMessage({
          type: "MovePreview",
          data: playedTiles.map((el) => pick(el, ["x", "y"])),
        });
    },
    [G.preview, canBeClicked, playerTiles, selectedTile, sendChatMessage]
  );

  const extraPlayerContent = useCallback(
    ({ tilesCount }) => (
      <>
        {t("game.info.tiles.owned")}: {tilesCount}
      </>
    ),
    [t]
  );

  return (
    <div ref={stickyRef}>
      <GameLayout
        gameName={t("game.name")}
        header={
          <div style={{ marginBottom: "14px" }}>
            <Header as="h2" textAlign="center">
              {t(`header.${phase}.${stage}`)}
              <Header.Subheader>{t(`subheader.${phase}.${stage}`)}</Header.Subheader>
            </Header>
            <Sticky context={stickyRef}>
              <TilesPanel
                disabled={!canMakeMove}
                onSwapTiles={() => setSwapTilesModal(true)}
                onSkipTurn={() => setSkipTurnModal(true)}
                tilesLeft={G.tilesLeft}
                {...{
                  playerTiles,
                  selectTile,
                  selectedTile,
                  onReturnTiles,
                  onPlayTiles,
                  canReturnTiles,
                  canSwapTiles,
                  canPlayTiles,
                }}
              />
            </Sticky>
          </div>
        }
        extraPlayerContent={extraPlayerContent}
      >
        <InfoPopup
          open={popupOpen && !selectedBlankTile && !swapTilesModal}
          errors={moveErrors}
          newWords={newWords(G.board, playerTiles)}
          specialBonus={playedTiles.length === 7}
          {...{ popupHandleRef }}
        />
        <Grid
          board={G.board}
          clickable={canBeClicked}
          handleFieldClick={clickBoard}
          playerTiles={stage === "wait_for_approval" ? G.pendingTiles : previewTiles || playerTiles}
          selectionEnabled={selectedTile !== null}
        />
        {G.pendingTiles && approvalPhase && (
          <ApprovalModal
            newWords={newWords(G.board, G.pendingTiles)}
            onReject={() => Approve(false)}
            onApprove={() => Approve(true)}
          />
        )}
        {swapTilesModal && (
          <SwapTilesModal
            onClose={() => {
              setSelectedForSwap([]);
              setSwapTilesModal(false);
            }}
            onSwap={() => {
              SwapTiles(selectedForSwap);
              setSwapTilesModal(false);
            }}
            {...{ playerTiles, onSelectForSwap, selectedForSwap }}
          />
        )}
        {selectedBlankTile !== null && (
          <BlankTileModal
            letters={currentLanguage.letters}
            onClose={() => {
              selectedBlankTile.replacement = blankTileReplacement;
              setPlayerTiles([...playerTiles]);
              setSelectedBlankTile(null);
            }}
            {...{ setBlankTileReplacement, blankTileReplacement }}
          />
        )}
        <Confirm
          open={skipTurnModal}
          onCancel={() => setSkipTurnModal(false)}
          onConfirm={onSkipTurn}
          cancelButton={t("game.modals.skip.cancel")}
          confirmButton={t("game.modals.skip.confirm")}
          content={t("game.modals.skip.content")}
        />
      </GameLayout>
    </div>
  );
};

export default GameBoard;
