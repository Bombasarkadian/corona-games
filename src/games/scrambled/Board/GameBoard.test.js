import React from "react";
import renderer from "react-test-renderer";
import GameBoard from "./GameBoard";
import BoardGameContextMock from "../GameContextMock";
import { MemoryRouter } from "react-router";

describe("scrambled GameBoard component", () => {
  it("renders correctly", () => {
    expect.hasAssertions();
    let component;
    renderer.act(() => {
      component = renderer.create(
        <MemoryRouter>
          <BoardGameContextMock>
            <GameBoard />
          </BoardGameContextMock>
        </MemoryRouter>
      );
    });
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
