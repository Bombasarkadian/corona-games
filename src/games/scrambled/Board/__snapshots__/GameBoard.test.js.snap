// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scrambled GameBoard component renders correctly 1`] = `
<div>
  <div
    className="ui container"
  >
    <div
      className="mainMenu"
    >
      <div
        className="ui container"
      >
        <a
          className="logo"
          href="/"
          onClick={[Function]}
        >
          <img
            className="ui image"
            slt="Octoboard logo"
            src="/images/octoboard-small.svg"
          />
        </a>
        <span
          className="gameName"
        >
          game.name
        </span>
        <span
          className="ui small label"
          onClick={[Function]}
          style={Object {}}
        >
          <i
            aria-hidden="true"
            className="open lock icon"
            onClick={[Function]}
          />
          <div
            className="detail"
          >
            game.public
          </div>
        </span>
      </div>
    </div>
    <div
      className="ui container"
    >
      <div
        style={
          Object {
            "marginBottom": "14px",
          }
        }
      >
        <h2
          className="ui center aligned header"
        >
          header.play.play
          <div
            className="sub header"
          >
            subheader.play.play
          </div>
        </h2>
        <div
          className=""
        >
          <div
            style={Object {}}
          />
          <div
            className="ui sticky"
          >
            <div
              className="ui segment"
            >
              <span
                className="ui green ribbon label"
                onClick={[Function]}
              >
                <i
                  aria-hidden="true"
                  className="star icon"
                  onClick={[Function]}
                />
                panel.tiles
              </span>
              <div
                className="ui top right attached label"
                onClick={[Function]}
              >
                game.info.tiles.left
                : 
                13
              </div>
              <div
                style={
                  Object {
                    "display": "flex",
                    "paddingTop": 10,
                  }
                }
              >
                <div
                  className="tile tile_separate tile_clickable"
                  onClick={[Function]}
                >
                  <div
                    className="letter"
                  >
                    A
                  </div>
                  <div
                    className="points"
                  >
                    1
                  </div>
                </div>
                <div
                  className="tile tile_separate tile_clickable"
                  onClick={[Function]}
                >
                  <div
                    className="letter"
                  >
                    B
                  </div>
                  <div
                    className="points"
                  >
                    2
                  </div>
                </div>
                <div
                  className="tile tile_separate tile_clickable"
                  onClick={[Function]}
                >
                  <div
                    className="letter letter_replacement"
                  />
                  <div
                    className="points"
                  >
                    0
                  </div>
                </div>
                <div
                  className="tile tile_separate tile_clickable"
                  onClick={[Function]}
                >
                  <div
                    className="letter"
                  >
                    Ż
                  </div>
                  <div
                    className="points"
                  >
                    9
                  </div>
                </div>
                <div
                  className="tile tile_separate tile_clickable"
                  onClick={[Function]}
                >
                  <div
                    className="letter"
                  >
                    Ó
                  </div>
                  <div
                    className="points"
                  >
                    7
                  </div>
                </div>
                <button
                  className="ui red disabled button"
                  disabled={true}
                  onClick={[Function]}
                  tabIndex={-1}
                >
                  <i
                    aria-hidden="true"
                    className="undo icon"
                    onClick={[Function]}
                  />
                  panel.actions.return
                </button>
                <button
                  className="ui orange button"
                  onClick={[Function]}
                >
                  <i
                    aria-hidden="true"
                    className="refresh icon"
                    onClick={[Function]}
                  />
                  panel.actions.swap
                </button>
                <button
                  className="ui yellow button"
                  onClick={[Function]}
                >
                  <i
                    aria-hidden="true"
                    className="forward icon"
                    onClick={[Function]}
                  />
                  panel.actions.skip
                </button>
                <button
                  className="ui green disabled button"
                  disabled={true}
                  onClick={[Function]}
                  tabIndex={-1}
                >
                  <i
                    aria-hidden="true"
                    className="check icon"
                    onClick={[Function]}
                  />
                  panel.actions.play
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="ui grid"
    >
      <div
        className="twelve wide computer sixteen wide mobile column"
      >
        <div
          className="row"
        >
          <div
            className="grid"
          >
            <table>
              <tbody>
                <tr
                  style={
                    Object {
                      "height": 46,
                    }
                  }
                >
                  <td
                    className="cell"
                  >
                    <div
                      className="field"
                      onClick={[Function]}
                      style={
                        Object {
                          "background": "#BB5C66",
                        }
                      }
                    >
                      <div
                        className="bonus"
                      >
                        <div
                          className="bonus_multiply"
                        >
                          3
                          x
                        </div>
                        <div
                          className="bonus_type"
                        >
                          game.info.bonuses.word
                        </div>
                      </div>
                      <div
                        className="field dimmer"
                      />
                    </div>
                  </td>
                  <td
                    className="cell"
                  >
                    <div
                      className="field"
                    >
                      <div
                        className="tile"
                      >
                        <div
                          className="letter"
                        >
                          L
                        </div>
                        <div
                          className="points"
                        >
                          2
                        </div>
                      </div>
                    </div>
                  </td>
                  <td
                    className="cell"
                  >
                    <div
                      className="field"
                    >
                      <div
                        className="tile"
                      >
                        <div
                          className="letter"
                        >
                          O
                        </div>
                        <div
                          className="points"
                        >
                          1
                        </div>
                      </div>
                    </div>
                  </td>
                  <td
                    className="cell"
                  >
                    <div
                      className="field"
                    >
                      <div
                        className="tile"
                      >
                        <div
                          className="letter letter_replacement"
                        >
                          L
                        </div>
                        <div
                          className="points"
                        >
                          0
                        </div>
                      </div>
                    </div>
                  </td>
                </tr>
                <tr
                  style={
                    Object {
                      "height": 46,
                    }
                  }
                >
                  <td
                    className="cell"
                  >
                    <div
                      className="field"
                      onClick={[Function]}
                    >
                      <div
                        className="field dimmer"
                      />
                    </div>
                  </td>
                  <td
                    className="cell"
                  >
                    <div
                      className="field"
                      onClick={[Function]}
                      style={
                        Object {
                          "background": "#21C4D8",
                        }
                      }
                    >
                      <div
                        className="bonus"
                      >
                        <div
                          className="bonus_multiply"
                        >
                          2
                          x
                        </div>
                        <div
                          className="bonus_type"
                        >
                          game.info.bonuses.letter
                        </div>
                      </div>
                      <div
                        className="field dimmer"
                      />
                    </div>
                  </td>
                  <td
                    className="cell"
                  >
                    <div
                      className="field"
                      onClick={[Function]}
                    >
                      <div
                        className="field dimmer"
                      />
                    </div>
                  </td>
                  <td
                    className="cell"
                  >
                    <div
                      className="field"
                      onClick={[Function]}
                    >
                      <div
                        className="field dimmer"
                      />
                    </div>
                  </td>
                </tr>
                <tr
                  style={
                    Object {
                      "height": 46,
                    }
                  }
                >
                  <td
                    className="cell"
                  >
                    <div
                      className="field"
                      onClick={[Function]}
                    >
                      <div
                        className="field dimmer"
                      />
                    </div>
                  </td>
                  <td
                    className="cell"
                  >
                    <div
                      className="field"
                      onClick={[Function]}
                      style={
                        Object {
                          "background": "#16919F",
                        }
                      }
                    >
                      <div
                        className="bonus"
                      >
                        <div
                          className="bonus_multiply"
                        >
                          3
                          x
                        </div>
                        <div
                          className="bonus_type"
                        >
                          game.info.bonuses.letter
                        </div>
                      </div>
                      <div
                        className="field dimmer"
                      />
                    </div>
                  </td>
                  <td
                    className="cell"
                  >
                    <div
                      className="field"
                      onClick={[Function]}
                      style={
                        Object {
                          "background": "#F9C54A",
                        }
                      }
                    >
                      <div
                        className="bonus"
                      >
                        <div
                          className="bonus_multiply"
                        >
                          2
                          x
                        </div>
                        <div
                          className="bonus_type"
                        >
                          game.info.bonuses.word
                        </div>
                      </div>
                      <div
                        className="field dimmer"
                      />
                    </div>
                  </td>
                  <td
                    className="cell"
                  >
                    <div
                      className="field"
                      onClick={[Function]}
                    >
                      <div
                        className="field dimmer"
                      />
                    </div>
                  </td>
                </tr>
                <tr
                  style={
                    Object {
                      "height": 46,
                    }
                  }
                >
                  <td
                    className="cell"
                  >
                    <div
                      className="field"
                      onClick={[Function]}
                      style={
                        Object {
                          "background": "#F9C54A",
                        }
                      }
                    >
                      <div
                        className="bonus"
                      >
                        <div
                          className="bonus_multiply"
                        >
                          2
                          x
                        </div>
                        <div
                          className="bonus_type"
                        >
                          game.info.bonuses.word
                        </div>
                      </div>
                      <div
                        className="field dimmer"
                      />
                    </div>
                  </td>
                  <td
                    className="cell"
                  >
                    <div
                      className="field"
                      onClick={[Function]}
                    >
                      <div
                        className="field dimmer"
                      />
                    </div>
                  </td>
                  <td
                    className="cell"
                  >
                    <div
                      className="field"
                      onClick={[Function]}
                    >
                      <div
                        className="field dimmer"
                      />
                    </div>
                  </td>
                  <td
                    className="cell"
                  >
                    <div
                      className="field"
                      onClick={[Function]}
                      style={
                        Object {
                          "background": "#F9C54A",
                        }
                      }
                    >
                      <i
                        aria-hidden="true"
                        className="star big disabled fitted icon"
                        onClick={[Function]}
                      />
                      <div
                        className="field dimmer"
                      />
                    </div>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
      <div
        className="four wide computer sixteen wide mobile column"
      >
        <div
          className="row"
        >
          <h2
            className="ui center aligned header"
            style={
              Object {
                "marginBottom": "-5px",
              }
            }
          >
            game.players
            <span
              className="ui small label"
              onClick={[Function]}
              style={Object {}}
            >
              <i
                aria-hidden="true"
                className="open lock icon"
                onClick={[Function]}
              />
              <div
                className="detail"
              >
                game.public
              </div>
            </span>
          </h2>
          <div
            className="ui segments"
          >
            <div
              className="ui segment player smoothDisabled"
            >
              <div
                className="ui feed"
              >
                <div
                  className="event"
                >
                  <div
                    className="label"
                  >
                    <div
                      style={
                        Object {
                          "display": "inline-block",
                          "height": "100%",
                          "verticalAlign": "middle",
                          "width": "100%",
                        }
                      }
                    >
                      <svg
                        className="octopus avatar"
                        style={
                          Object {
                            "fill": "#000",
                            "stroke": "#000",
                            "strokeLinecap": "round",
                            "strokeWidth": 0,
                            "width": "100%",
                          }
                        }
                        viewBox="0 0 200 200"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <g
                          style={
                            Object {
                              "fill": "#f4d0b5",
                              "stroke": "#f4d0b5",
                            }
                          }
                        >
                          <svg>
                            03.svg
                          </svg>
                        </g>
                        <svg>
                          02.svg
                        </svg>
                        <svg>
                          empty.svg
                        </svg>
                        <svg>
                          02.svg
                        </svg>
                        <svg>
                          04.svg
                        </svg>
                      </svg>
                    </div>
                  </div>
                  <div
                    className="content"
                  >
                    <div
                      className="date"
                    >
                      user-0
                       
                      <span>
                        (
                        sidebar.player.you
                        )
                      </span>
                       
                      <i
                        aria-hidden="true"
                        className="green check circle icon"
                        onClick={[Function]}
                      />
                      <i
                        aria-hidden="true"
                        className="broken chain icon smoothDisabled hidden"
                        onClick={[Function]}
                        style={
                          Object {
                            "opacity": 0,
                          }
                        }
                      />
                    </div>
                    <div
                      className="content"
                    >
                      <i
                        aria-hidden="true"
                        className="yellow trophy icon"
                        onClick={[Function]}
                      />
                      sidebar.player.points
                    </div>
                    <div
                      className="text extra extra"
                    >
                      <div
                        className="ui middle aligned list"
                        role="list"
                      />
                    </div>
                    <div
                      className="text extra extra"
                    >
                      game.info.tiles.owned
                      : 
                      5
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              className="ui disabled segment player smoothDisabled"
            >
              <div
                className="ui feed"
              >
                <div
                  className="event"
                >
                  <div
                    className="label"
                  >
                    <div
                      style={
                        Object {
                          "display": "inline-block",
                          "height": "100%",
                          "verticalAlign": "middle",
                          "width": "100%",
                        }
                      }
                    >
                      <svg
                        className="octopus avatar"
                        style={
                          Object {
                            "fill": "#000",
                            "stroke": "#000",
                            "strokeLinecap": "round",
                            "strokeWidth": 0,
                            "width": "100%",
                          }
                        }
                        viewBox="0 0 200 200"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <g
                          style={
                            Object {
                              "fill": "#a5673f",
                              "fillOpacity": 0.3,
                              "stroke": "#a5673f",
                              "strokeWidth": 5,
                            }
                          }
                        >
                          <svg>
                            01.svg
                          </svg>
                        </g>
                        <svg>
                          03.svg
                        </svg>
                        <svg>
                          01.svg
                        </svg>
                        <svg>
                          03.svg
                        </svg>
                        <svg>
                          05.svg
                        </svg>
                      </svg>
                    </div>
                  </div>
                  <div
                    className="content"
                  >
                    <div
                      className="date"
                    >
                      user-1
                       
                       
                      <i
                        aria-hidden="true"
                        className="broken chain icon smoothDisabled "
                        onClick={[Function]}
                        style={
                          Object {
                            "opacity": 0.45,
                          }
                        }
                      />
                    </div>
                    <div
                      className="content"
                    >
                      <i
                        aria-hidden="true"
                        className="grey trophy icon"
                        onClick={[Function]}
                      />
                      sidebar.player.points
                    </div>
                    <div
                      className="text extra extra"
                    >
                      <div
                        className="ui middle aligned list"
                        role="list"
                      />
                    </div>
                    <div
                      className="text extra extra"
                    >
                      game.info.tiles.owned
                      : 
                      7
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            className="ui basic center aligned segment leaveSegment"
          >
            <button
              className="ui blue icon disabled left labeled button"
              disabled={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              tabIndex={-1}
            >
              <i
                aria-hidden="true"
                className="info circle icon"
                onClick={[Function]}
              />
              game.rules.header
            </button>
          </div>
          <div
            className="ui basic center aligned segment leaveSegment"
          >
            <button
              className="ui red icon left labeled button"
              onClick={[Function]}
            >
              <i
                aria-hidden="true"
                className="close icon"
                onClick={[Function]}
              />
              game.leave.button
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      className="wrapper"
    >
      <span
        className="triggerButton"
        data-unread={0}
        onClick={[Function]}
      />
      <div
        className="popupChat"
      >
        <div
          className="ui segments chat"
        >
          <div
            className="ui compact segment header"
          >
            game.chat.header
            <span
              className="closeIcon"
              onClick={[Function]}
            >
              <i
                aria-hidden="true"
                className="close icon"
                onClick={[Function]}
              />
            </span>
          </div>
          <div
            className="ui segment messages"
          >
            <div
              className="ui comments"
            />
          </div>
          <div
            className="ui segment inputBox"
          >
            <form
              onSubmit={[Function]}
            >
              <div
                className="ui fluid icon input"
              >
                <input
                  onChange={[Function]}
                  placeholder="game.chat.sendMessage"
                  type="text"
                  value=""
                />
                <i
                  aria-hidden="true"
                  className="send icon"
                  onClick={[Function]}
                />
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
    <div
      className="Toastify"
    />
  </div>
</div>
`;
